tuist/tuist
# Fix errors when archiving projects with static XCFrameworks

Resolves: https://github.com/tuist/tuist/issues/5152

### Short description 📝

- Archiving projects with static XCFramework dependencies can yield the following error

```
Multiple commands produce: ....
```

- This issue was accidentally introduced while attempting to add better static XCFramework support in https://github.com/tuist/tuist/pull/5107
- It is caused due to the technique employed to get Xcode to process the XCFramework (and thereby have its symbols visible to the target) without explicitly linking it to avoid duplicate symbol issues that can arise when having static targets depend on static products.
  - The XCFramework as added to a custom Copy Files build phase that would copy it to the products directory
  - The same technique is currently employed by regular static targets, where Xcode simply ignores the copy if the product already exists
  - This didn't seem to work with XCFrameworks
  - The custom copy phase is marked with "Copy only when installing" which is why the issue surfaced when archiving only
- There are a few options to resolve this (see https://github.com/tuist/tuist/issues/5152)
- A solution that is closer to the current implementation involves specifying unique destinations per target in the custom copy phase for XCFrameworks to avoid the conflict
  - This is somewhat redundant / wasteful as those copies won't be used and only serve as way for Xcode to process the XCFramework without explicitly linking it nor produce the build error due to multiple targets having the same build identical build action
- A new copy files build phase is added for this
- The fixture has now been extended to include more targets and the acceptance test to perform an archive to verify the new solution


### How to test the changes locally 🧐

- Build this version of tuist

```
swift build
```

- Generate the fixture `ios_app_with_xcframeworks`

```
swift run tuist generate --path projects/tuist/fixtures/ios_app_with_xcframeworks
```

- Inspect the generated project
- Verify the new copy files build phases (named "Static XCFramework Dependencies") are added to `StaticFrameworkA` and `StaticFrameworkB`
- Veirfy the following targets build successfully in isolation (perform a clean + select the appropriate scheme)
  - `App
  - `StaticFrameworkA`
  - `StaticFrameworkB`
- Verify archiving `App` succeeds (note: may need to temporarily select a development team to get this working)

### Contributor checklist ✅


### Reviewer checklist ✅
